pyclesperanto_prototype/_tier3/_morphological_chan_vese.py

```diff
@@ -0,0 +1,170 @@
+    borders. It is required that the inside of the object looks different on
+    average than the outside (i.e., the inner area of the object should be
+    darker or lighter than the outer area on average).
+
+    This is a re-implementation of the algorithm in scikit-image [2].
+
+    Parameters
+    ----------
+    input_image: Image
+    contour_image: Image, optional
+    output_image: Image, optional
+    num_iter : uint
+        Number of iterations to run
+    smoothing: int, optional
+    lambda1 : float, optional
+        Weight parameter for the outer region. If `lambda1` is larger than
+        `lambda2`, the outer region will contain a larger range of values than
+        the inner region.
+    lambda2 : float, optional
+        Weight parameter for the inner region. If `lambda2` is larger than
+        `lambda1`, the inner region will contain a larger range of values than
+        the outer region.
+
+    Returns
+    -------
+    Final segmentation
+
+    Notes
+    -----
+    This is a version of the Chan-Vese algorithm that uses morphological
+    operators instead of solving a partial differential equation (PDE) for the
+    evolution of the contour. The set of morphological operators used in this
+    algorithm are proved to be infinitesimally equivalent to the Chan-Vese PDE
+    (see [1]_). However, morphological operators are do not suffer from the
+    numerical stability issues typically found in PDEs (it is not necessary to
+    find the right time step for the evolution), and are computationally
+    faster.
+
+    The algorithm and its theoretical derivation are described in [1]_.
+
+    References
+    ----------
+    .. [1] A Morphological Approach to Curvature-based Evolution of Curves and
+           Surfaces, Pablo Márquez-Neila, Luis Baumela, Luis Álvarez. In IEEE
+           Transactions on Pattern Analysis and Machine Intelligence (PAMI),
+           2014, :DOI:`10.1109/TPAMI.2013.106`
+    .. [2] https://github.com/scikit-image/scikit-image/blob/5e74a4a3a5149a8a14566b81a32bb15499aa3857/skimage/segmentation/morphsnakes.py#L212-L312
+    """
+    from .._tier0 import create_like
+    from .._tier1 import superior_inferior, inferior_superior, set
+
+    from .._tier1 import absolute, binary_or, binary_not, gradient_x, gradient_y
+    from .._tier1 import gradient_z, copy, power
+    from .._tier1 import greater_constant, smaller_constant
+    from .._tier1 import mask, add_image_and_scalar, add_images_weighted
+
+    if contour_image.size == 0:
+        contour_image = checkerboard_level_set(input_image.shape)
+    
+    greater_constant(contour_image, destination=output_image, constant=0)
+
+    temp_1 = create_like(output_image)
+    temp_2 = create_like(output_image)
+    temp_3 = create_like(output_image)
+    temp_4 = create_like(output_image)
+    temp_5 = create_like(output_image)
+
+    for _ in range(num_iter):
+        set(temp_3, 0)
```

Comment:
Hint: I had to initialize this image here with 0 because the `create*` functions only allocate memory. Further down, we compute `temp_3 += ....` which might be dangerous with a non-initialized image.